crates/libcgroups/src/v1/freezer.rs

```diff
@@ -114,17 +122,43 @@ impl Freezer {
         Ok(())
     }
 
-    fn read_freezer_state(cgroup_root: &Path) -> Result<String, WrappedIoError> {
+    pub fn read_freezer_state(cgroup_root: &Path) -> FreezerState {
         let path = cgroup_root.join(CGROUP_FREEZER_STATE);
-        let mut content = String::new();
-        OpenOptions::new()
-            .create(false)
-            .read(true)
-            .open(path)
-            .wrap_open(cgroup_root)?
-            .read_to_string(&mut content)
-            .wrap_read(cgroup_root)?;
-        Ok(content)
+        loop {
+            let state = match read_to_string(&path) {
+                Ok(state) => state,
+                Err(_) => {
+                    return FreezerState::Undefined;
+                }
+            };
+            match state.trim() {
+                FREEZER_STATE_THAWED => return FreezerState::Thawed,
+                FREEZER_STATE_FROZEN => {
+                    // Find out whether the cgroup is frozen directly,
+                    // or indirectly via an ancestor.
+                    let self_state = match read_to_string(path.join(CGROUP_FREEZER_SELF_STATE)) {
+                        Ok(self_state) => self_state,
+                        Err(_) => {
+                            return FreezerState::Frozen;
+                        }
+                    };
+                    return match self_state.as_str() {
+                        "0\n" => FreezerState::Thawed,
+                        "1\n" => FreezerState::Frozen,
+                        _ => FreezerState::Undefined,
+                    };
+                }
```

Comment:
I'm not sure this is what we should follow. From kernel docs
```
The cgroup freezer is hierarchical. Freezing a cgroup freezes all tasks belonging to 
the cgroup and all its descendant cgroups. Each cgroup has its own state (self-state) 
and the state inherited from the parent (parent-state). Iff both states are THAWED, 
the cgroup is THAWED.
```
https://docs.kernel.org/admin-guide/cgroup-v1/freezer-subsystem.html

So, if we are getting frozen, we should return frozen even if self_state is thawed, right?